lua/lean/_util.lua

```diff
@@ -23,4 +24,46 @@ function M.set_augroup(name, autocmds, buffer)
   ]], name, name, buffer_string, autocmds))
 end
 
+--- Run a subprocess, blocking on exit, and returning its stdout.
+---
+--- Unlike `system()`, we don't mix stdout and stderr, and unlike
+--- `vim.loop.spawn`, we wait for process exit and collect the output.
+--- @return table: the lines of stdout of the exited process
+function M.subprocess_check_output(cmd, opts, timeout)
+  if not timeout then timeout = 5000 end
+  local lines, stderr_lines
+
+  opts = vim.tbl_extend(
+    "keep", opts or {}, {
+      stdout_buffered = true,
+      stderr_buffered = true,
+      on_stdout = function(_, data) lines = data end,
+      on_stderr = function(_, data) stderr_lines = data end
+    }
+  )
```

Comment:
[Plenary's `Job` module](https://github.com/nvim-lua/plenary.nvim/blob/master/lua/plenary/job.lua) abstracts away a lot of these lower level details, perhaps there's an argument for using that instead of "manually" spawning a process with handlers?

Esp. relevant here are [`Job:result()`](https://github.com/nvim-lua/plenary.nvim/blob/8bae2c1fadc9ed5bfcfb5ecbd0c0c4d7d40cb974/lua/plenary/job.lua#L453), [`Job:stderr_result()`](https://github.com/nvim-lua/plenary.nvim/blob/8bae2c1fadc9ed5bfcfb5ecbd0c0c4d7d40cb974/lua/plenary/job.lua#L458), and [`Job:wait()`](https://github.com/nvim-lua/plenary.nvim/blob/8bae2c1fadc9ed5bfcfb5ecbd0c0c4d7d40cb974/lua/plenary/job.lua#L468).

Reply:
It'd add a runtime dep for users (instead of just a testing one).

It's definitely annoying to need to NIH this though, I was complaining about it a bit.

But happy to go either way.

Reply:
Plenary also has a nice async library with wrappers for LSP that would be nice to use instead of the callback hell.

Reply:
Good point, IMO since our user-base seems like it still is pretty much still just us (and will be until we implement competitive features like #84) I'd say it's best to add the dependency now - I can work later today on refactoring this to use `Job`.

Reply:
OK, sounds good to me! And thanks that'd be great, gonna probably hit a beer for the evening so feel free to play around, otherwise I'll address in the morning.